### ISSUE TITLE
Better way to define workflows

### ISSUE THREAD

**Issue by SomeKittens** [SomeKittens/gustav#14]
Currently (with manually assigned ids) it's very difficult to define a workflow.

Possible solution is to define `createNode` (again) with a different signature:

```typescript
createNode (name: string, config?: Object, ...dependencies:NodeSymbol)
```

Config may not be an object either.

**Comment by SomeKittens** [comment]
# Proposal:

Add a function `createWorkflow` that generates an empty workflow.

```typescript
let wf = gustav.createWorkflow();
```

(In all following examples, you can assume `wf` is initialized as above unless otherwise overridden in the example).

`Workflow` now exposes several new methods:

 - `.source`
 - `.transf`
 - `.sink`

Each of these adds the given type of node to the graph.  When called on a node (rather than the root workflow), the newly created node will depend on the node it was called from.  `.source` and `.transf` return the node they generate so that dependency chains are easily visible.

Simple example of [this](https://github.com/SomeKittens/gustav/blob/master/test/samples.ts#L24) workflow:

```typescript
wf.source('intSource')
  .transf('timesTwo')
  .sink('fromIntTransformer');
```

The `.sink` method would return the workflow itself, enabling the following:

```typescript
let wf = gustav.createWorkflow()
  .source('intSource')
  .transf('timesTwo')
  .sink('fromIntTransformer');
```

## Forking & merging

So far this has just covered straight-line workflows.  The case of forking a stream is fairly simple:

```typescript
let sauce = wf.source('intSource');

sauce.sink('fromIntSource');
sauce.transf('timesTwo')
     .sink('fromIntTransformer');
```

Merging is simply another method:

```typescript
let twits = wf.source('twits', {
  search: 'SomeKittens'
});
let fbs = wf.source('fbs', {
  search: 'Kerbal Space Program'
});

twits
  .merge(fbs)
  .transf('sentiment')
  .sink('socket');
```

`.merge(...streams)` returns a new type of node, one that is the result of combining multiple streams.

## Forking redux

Sometimes we just want to add a sink at some point in the stream (for benchmarking or metadata purposes).  Thus we have a special method for that: `.singleSink`.  This method returns the previous node (instead of the workflow), allowing for this:

```typescript
let wf = gustav.createWorkflow()
  .source('kafkaSource', {topic: 'bench-emails'})
  .singleSink('benchmark', {blah: blah0}) // this returns the source
  .transf('getUrls')
  .singleSink('benchmark', {blah: blah0}) // reutrns getUrls
  .sink('kafkaOut', {topic: 'bench-urls'}); // returns the workflow proper
```

(still not sold on the name `singleSink`)

**Comment by SomeKittens** [comment]
Legacy JSON definition of workflows should still be supported so that it's possible to pass a defined workflow between processes, server/browser, etc.

**Comment by SomeKittens** [comment]
`s/singleSink/tap/`

**Comment by SomeKittens** [comment]
Landed in 66094fbc4cb1db8e5f96facc3899ac666a1b818a
